### ISSUE TITLE
Proposal for contributing WatermelonDB Migration Generator 

### ISSUE THREAD

**Issue by 509dave16** [Nozbe/WatermelonDB#764]
Hey, @radex !

I previously made a migration yeoman generator for `realm-js` to help with managing [linear migrations](https://realm.io/docs/javascript/latest/#linear-migrations): https://www.npmjs.com/package/generator-realm-migration

WatermelonDB's migrations are understandably different from `realm-js`. But the same core scenario occurs where some table's schema gets changed and you need a migration created for it. This migration wouldn't cover performing CRUD operations to to migrate existing data to the next schema version. But it would at least cover the creation of the migration steps that would mirror the changes you made to the database schema.

What would be all of the schema alterations we would need to cover?
- Create Table
- Drop Table
- Alter Table
  - Add Column
  - ???

I'm wondering if you would first like help to complete the supported schema operation steps, before even proceeding forward with a migration generator based on your comment here: https://nozbe.github.io/WatermelonDB/Advanced/Migrations.html#migration-steps

In any case, let me know what you think!

**Comment by radex** [comment]
@509dave16 A migration generator would be an awesome contribution to the project! I'd love that — as I'm sure, many users would :)

Currently, there's just two types of migrations: create table, add columns (to existing table). 

> I'm wondering if you would first like help to complete the supported schema operation steps

Sure, that would be quite nice. A complication stems from the fact that the migrations have to be performed on both SQLite and LokiJS. Although a contribution that only implements the migration on one of the platforms is acceptable.

Destroying tables seems easy. Adding/removing indices also seems easy. Renaming or deleting columns, if I recall correctly, is tricky with SQLite.  It would also require an expensive linear scan and full save on LokiJS (but seems easy to implement, just a forEach and a update on all records in a collection).

**Comment by 509dave16** [comment]
@radex Now that I think about it. It might be best to go ahead and just start work on the migration generator now. And then start adding support for more migration steps which would need to be added to the generator.

Would you prefer if this was just a community package published by me to npm to start? Which is basically what I did for: https://www.npmjs.com/package/generator-realm-migration

Also we'll need to decide what file structure(s) to support. The previous generator I made used a static `schema` property off of each Model. And then compared them to the most recent archived schema version.

I think I need to look at `appSchema`, `tableSchema`, `createTable`, and `addColumns` more. Basically what I plan on doing for WatermelonDB schemas is the following:
1. Derive table schema's from migrations
2. Compare derived table schema's with declared table schemas to determine the difference.
3. Create migration that add's steps for the difference.

Basically at the beginning if there's no migrations that introduce a table, then we know that the new migration needs to add a `createTable` step. And then if there's new columns for a table schema then we just need to add a `addColumns` step.

**Comment by radex** [comment]
> Would you prefer if this was just a community package published by me to npm to start? Which is basically what I did for: https://www.npmjs.com/package/generator-realm-migration

Yup, that seems easiest to manage - and then we could pull that into the mainline, or just add a dependency, depending on what's more convenient for everyone

> Derive table schema's from migrations
> Compare derived table schema's with declared table schemas to determine the difference.
> Create migration that add's steps for the difference.

That sounds quite reasonable. But why not the other way around? Write a migration, and then migrate the schema based on that.

**Comment by 509dave16** [comment]
@radex You're right! I completely forgot about how WatermelonDB's migrations are incremental just like those of Eloquent, Sequelize, Knex, etc... . All together they describe the database schema. Then we are only ever having to create/update one file.

**Comment by radex** [comment]
> All together they describe the database schema

Not necessarily. Migrations don't _have to_ describe how to get from nothing to a full schema. you can start from an initial schema and only then start supporting migrations ;/

**Comment by 509dave16** [comment]
@radex Oh that's right. It's opt in. Hmm.... . Got it. So basically the migrations aren't guaranteed to fully describe the database schema because they can start at a future point in time of the schema. Which is fine because all we really care about is what's in the migrations that's not in the database schema definition. In which case like you mentioned previously, we could update the database schema definition with what is new from the migrations.

**Comment by radex** [comment]
> In which case like you mentioned previously, we could update the database schema definition with what is new from the migrations.

Yup! It seems more than doable since the schema definition contains a "version", and a migration also specifies which version it migrates to (and it's always +1 from the previous version). So you can very easily figure out which migration definitions need to be applied to schema.

**Comment by pokey** [comment]
Fwiw, Alembic has [support](https://alembic.sqlalchemy.org/en/latest/batch.html) for dropping / altering columns in SQLite tables by bulk recreating the table, and will even automatically generate a migration to do so.  Might be worth having a peek at their impl if Watermelondb decides to support such a thing down the road

**Comment by gregoryfm** [comment]
Hi, is there any plan to add more types of migrations, such as deleting or renaming tables and/or columns?

Can I help with anything?

Thanks

**Comment by radex** [comment]
@gregoryfm I don't have such plans at the moment, but would more than welcome contributions!

**Comment by landonwjohnson** [comment]
I would like to see a migration for sequlize and mysql.
